### DISCUSSION TITLE
"Manage applications" problem and the fix

### DISCUSSION THREAD

**Discussion by adelie93** [rgerganov/gen5fw#19] [General]
As I and others have posted in other threads, after using `security_force` method to modify firmware, many settings items change their wording to "Manage applications" instead of the proper text.

# Identifying the root cause
The "Manage applications" texts all appear (for me) within views from the "Settings" app.
We must look at the Settings app on the live system: `/system/app/Settings.apk`
Take a complete copy of the `system` partition from the eMMC onto a USB drive for further investigation:
```
# On the head unit, with a USB drive plugged in
cd /storage/usb0/
busybox dd if=/dev/block/mmcblk0p2 of=system.live.ext4 bs=1M
```
In theory, this should be identical to the `system.ext4` image that we upload with `security force` method...
However, this is not the case!
```
# On my local Linux system
mount ~/gen5fw/fw/system.ext4 /tmp/car
mount system.live.ext4 /tmp/car_live
diff -r /tmp/car /tmp/car_live
```
There are actually *many* differences between the two... to include `Settings.apk`!
Here is the output I got from my system:
```
Binary files /tmp/car/app/CalendarProvider.odex and /tmp/car_live/app/CalendarProvider.odex differ
Binary files /tmp/car/app/ContactsProvider.odex and /tmp/car_live/app/ContactsProvider.odex differ
Binary files /tmp/car/app/DownloadProvider.odex and /tmp/car_live/app/DownloadProvider.odex differ
Binary files /tmp/car/app/HKMC_EngineerMode.apk and /tmp/car_live/app/HKMC_EngineerMode.apk differ
Binary files /tmp/car/app/Launcher2.odex and /tmp/car_live/app/Launcher2.odex differ
Binary files /tmp/car/app/Settings.apk and /tmp/car_live/app/Settings.apk differ
Binary files /tmp/car/app/Settings.odex and /tmp/car_live/app/Settings.odex differ
Binary files /tmp/car/bin/perf and /tmp/car_live/bin/perf differ
Binary files /tmp/car/framework/framework-res.apk and /tmp/car_live/framework/framework-res.apk differ
Binary files /tmp/car/framework/framework-res.apk.ORIG and /tmp/car_live/framework/framework-res.apk.ORIG differ
Binary files /tmp/car/lib/libandroid_runtime.so and /tmp/car_live/lib/libandroid_runtime.so differ
Binary files /tmp/car/lib/libMzVrm.so and /tmp/car_live/lib/libMzVrm.so differ
Binary files /tmp/car/lib/libwebcore.so and /tmp/car_live/lib/libwebcore.so differ
Binary files /tmp/car/lib/modules/vpu_lib.ko and /tmp/car_live/lib/modules/vpu_lib.ko differ
```

I use `vbindiff` to examine the difference between the two:
![image](https://github.com/rgerganov/gen5fw/assets/109261315/7b6d04bf-793d-45da-b10c-18ada6189cef)

You can see that long continuous runs of `0xFF` have somehow been changed to `0x00`. This causes the app to behave unpredictably (possibly caused by Android app integrity checks failing to get the proper checksum on the file? Not sure)

I believe this to be due to the sparse image technique that we must use with `security force`, combined with possibly incorrect handling of uninitialized blocks by the kernel: see [this patch](https://kernel.googlesource.com/pub/scm/linux/kernel/git/stable/stable-queue/+/v3.4.3/releases/3.0.4/ext4-fix-nomblk_io_submit-option-so-it-correctly-converts-uninit-blocks.patch) for (a little) more information.

# Fixing the problem
So. After identifying this problem, I attempt to fix it:
```
# On the head unit, over ADB...
# remount the system partition so it can be changed
# note the use of nomblk_io_submit to address the problem of uninitialized blocks
mount -o rw,remount,nomblk_io_submit /system
cd /system/app
# Take a backup of the (broken) Settings apps in case something goes wrong
mv Settings.apk Settings.apk.ORIG
mv Settings.odex Settings.odex.ORIG
# copy the original (correct) app to the device
cp /storage/usb0/Settings.apk Settings.apk
cp /storage/usb0/Settings.odex Settings.odex
# Maybe not necessary but ensure that our writes are flushed to disk
sync
# Reboot to see if this fixes things
reboot
```

# Result
**Now my settings screens once again show the proper text instead of "Manage applications".**

Please let me know if this fix works for anyone else.

# Follow-up
I actually took this one step further and attempted a complete replacement of the `system` partition:
```
cd /storage/usb0
busybox dd if=system.ext4 of=/dev/block/mmcblk0p2 bs=1M
reboot
```
After doing this, then taking a new copy of the filesystem as detailed above, this actually fixed *most* (but not *all*) of the problems that `diff -r` showed.

This only fixes part of my problem...  I no longer get the "Manage applications" error, which is wonderful. But on some other screens I still experience the strange `Del <X]` icon that seems to indicate a missing proper icon. So I think there is still a bit more work to do.

**I think this is probably a good indicator of *why* the problem is occurring: an issue with uninitialized blocks on the eMMC.**
Anyone have any good suggestions how to fix that while still using `security force`?

**Comment by rgerganov** [comment]
Thank you for troubleshooting this problem, this is very helpful.

I wonder if we run into this if we flash unmodified ext4 partition converted to sparse with the `security_force` method. There is some chance that we corrupt something when we modify the partition to enable ADB.

Maybe we should mount with the `nomblk_io_submit` option when we do the changes, i.e.:
```
$ sudo mount -o nomblk_io_submit system.ext4 /tmp/car
$ vim /tmp/car/etc/permissions/com.hkmc.software.engineermode.adb_hide.xml
$ sudo umount /tmp/car
```

Can you try this out?

**Reply by adelie93** [reply]
I don't think we are corrupting the image. The errors happen even if I use security force to add in the original unmodified image. I think it must be some difference in the way the sparse image is written compared to the non-sparse image written by the normal upgrade process.

**Comment by rgerganov** [comment]
> But on some other screens I still experience the strange `Del <X]` icon that seems to indicate a missing proper icon. So I think there is still a bit more work to do.

Which screen is that? We can try to decompile the corresponding app and try to understand what the problem is.

**Reply by adelie93** [reply]
One annoying place it happened was on the volume adjustment. When I changed the volume with the knob, the volume bar at the bottom of the screen would have the weird missing icon `Del <X]` thing on it.

But. I fixed it!

My `/system` partition matched every file, so I wasn't sure what could possibly be missing...  Until I thought: Oh, what if it's using a cached copy from before I fixed the partition.
And that was the answer. The QuickBoot image was made at first boot, before I fixed the `/system` partition with the `dd`.
So, I blew away the QuickBoot image and let it rebuild, and everything is perfect now:
```
busybox dd if=/dev/zero of=/dev/block/mmcblk0p10 bs=1M
reboot
```
No more problems. Everything looks exactly as stock, except for the changes I wanted to make.

**Reply by rgerganov** [reply]
Thank you very much for sharing these fixes. Great work!

**Reply by syboxez** [reply]
> One annoying place it happened was on the volume adjustment. When I changed the volume with the knob, the volume bar at the bottom of the screen would have the weird missing icon `Del <X]` thing on it.
> 
> But. I fixed it!
> 
> My `/system` partition matched every file, so I wasn't sure what could possibly be missing... Until I thought: Oh, what if it's using a cached copy from before I fixed the partition. And that was the answer. The QuickBoot image was made at first boot, before I fixed the `/system` partition with the `dd`. So, I blew away the QuickBoot image and let it rebuild, and everything is perfect now:
> 
> ```
> busybox dd if=/dev/zero of=/dev/block/mmcblk0p10 bs=1M
> reboot
> ```
> 
> No more problems. Everything looks exactly as stock, except for the changes I wanted to make.

How were you able to find out which partition QuickBoot was on?

**Reply by syboxez** [reply]
> > One annoying place it happened was on the volume adjustment. When I changed the volume with the knob, the volume bar at the bottom of the screen would have the weird missing icon `Del <X]` thing on it.
> > But. I fixed it!
> > My `/system` partition matched every file, so I wasn't sure what could possibly be missing... Until I thought: Oh, what if it's using a cached copy from before I fixed the partition. And that was the answer. The QuickBoot image was made at first boot, before I fixed the `/system` partition with the `dd`. So, I blew away the QuickBoot image and let it rebuild, and everything is perfect now:
> > ```
> > busybox dd if=/dev/zero of=/dev/block/mmcblk0p10 bs=1M
> > reboot
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > No more problems. Everything looks exactly as stock, except for the changes I wanted to make.
> 
> How were you able to find out which partition QuickBoot was on?

To answer my own question, look at the kernel cmdline for a `resume` option via `cat /proc/cmdline`. My qb partition was also /dev/mmcblk0p10

**Comment by syboxez** [comment]
So I'm getting a bit of an issue. I flashed using security force, and I'm getting the "Manage applications" issue, along with the "Del" issue.

Problem is, now I can't select anything in my settings menu, so I can't get into engineering mode to enable ADB or even flash the original OTA update.

Any idea what I can do from here?

I was thinking that I could maybe inject a terminal emulator APK to gain root and restore the original Settings.apk and odex, but I'm not sure if I can just drop the apk into /system/app or if I need to create an odex file for the terminal app. Also not sure if I need a third party launcher to access the app. Also not sure if I need to inject an on screen keyboard or if it comes with one.

I have an ADPE

**Reply by syboxez** [reply]
I was able to inject ADW Launcher, get into Engineering mode, and enable adb. The terminal emulator I injected refused to launch.

EDIT: The $hell terminal emulator works, but trying to run /system/bin/amossu does not give me root. Is there an alternate way for me to have root? Do I need to use adb for it to work?

**Comment by dbolger** [comment]
I ran into this issue also after using security_force. I'm working on updating my vehicle using an SD card official patch currently to see if it reverts the breaking firmware.

**Reply by dbolger** [reply]
Worth adding: radio intermittently works when in this 'semi-bricked' state. Once you try to play music in CarPlay or SiriusXM, all audio related to music stops working entirely. Lane assist noises and buttons on the head unit still invoke 'dings' but audio is lost until you select the radio as your audio source, then restart the head unit.

**Reply by dbolger** [reply]
Navigation was also completely shot. My guess here is that since the Navigation Updater asked to update through the SD card, I should have went that route instead of updating with the patched firmware via USB. I've read stories of folks having issues with navigation after using a USB instead of the SD card to update. I think this depends on the year your car is. Some models seem to support USB while others don't.

**Reply by dbolger** [reply]
Also a potential factor: this vehicle hadn't had any software updates. I was running TM.USA.SOP.V115.200821.STD_H which I'm sure had some minor versions between it and the latest version. In hindsight I probably should have done a typical user update and then attempted to patch the firmware.

**Reply by dbolger** [reply]
Update: restoring from an unmodified version of the firmware using the SD card fixed the Del/Manage Applications issue as well.

**Reply by dbolger** [reply]
The mismatch between the actual head unit version and the navigation version stored on the SD card could have caused navigation to fail. I'm currently putting the latest navigation version on the original SD card with the vehicle and confirming. I will attempt to patch the firmware via USB once the navigation matches.

I suspect I'll get the Del/Manage Applications issue again, but I'd like to confirm the radio/CarPlay/SiriusXM still works after doing the custom patch via USB. If not, there may need to be a different route for patching. If it does still work, I'll be attempting the fix to see if it works in my scenario.

**Reply by dbolger** [reply]
It seems my firmware has been updated or is in a broken state where I can no longer upload custom firmware via USB. I get an error saying `=== RSA signature failed ===` immediately after entering the secret recovery mode. This is happening with and without a USB stick inserted (USB containing security_force) It's definitely coming out of lk.rom, and my worry here is that something in lk.rom may have changed that no longer allows for custom firmware to be installed (a mitigation perhaps).
